apis/metal3.io/v1alpha1/baremetalhost_validation.go

```diff
@@ -184,3 +256,95 @@ func validateImageURL(imageURL string) error {
 
 	return nil
 }
+
+// When making changes to this function for operationalstatus and errortype,
+// also make the corresponding changes in the OperationalStatus and
+// ErrorType fields in the struct definition of BareMetalHostStatus in
+// the file baremetalhost_types.go
+func checkStatusAnnotation(bmhstatus *BareMetalHostStatus) error {
+
+	operatstatus_allowed := []string{"", "OK", "discovered", "error", "delayed", "detached"}
+	errortype_allowed := []string{"", "provisioned registration error", "registration error", "inspection error", "preparation error", "provisioning error", "power management error"}
+
+	if !stringInSlice(operatstatus_allowed, string(bmhstatus.OperationalStatus)) {
+		return fmt.Errorf("invalid OperationalStatus='%s' in StatusAnnotation", string(bmhstatus.OperationalStatus))
+	}
+
+	if !stringInSlice(errortype_allowed, string(bmhstatus.ErrorType)) {
+		return fmt.Errorf("invalid ErrorType='%s' in StatusAnnotation", string(bmhstatus.ErrorType))
+	}
+
+	return nil
+}
+
+func validHwdDetailsAnnotation(hwdetannotaiton string, inspect string) error {
+	if hwdetannotaiton == "" {
+		return nil
+	}
+
+	if inspect != "disabled" {
+		return fmt.Errorf("inspection has to be disabled for HardwareDetailsAnnotation, check if {'inspect.metal3.io' : 'disabled'}")
+	}
+
+	objStatus := &HardwareDetails{}
+	err := json.Unmarshal([]byte(hwdetannotaiton), objStatus)
+	if err != nil {
+		return fmt.Errorf("failed to fetch Hardware Details from annotation, error=%w", err)
+	}
+
+	deco := json.NewDecoder(strings.NewReader(hwdetannotaiton))
+	deco.DisallowUnknownFields()
+	if err = deco.Decode(objStatus); err != nil {
+		return fmt.Errorf("invalid field in Hardware Details Annotation, error=%w", err)
+	}
+
+	return nil
+}
+
+func validInspectAnnotation(inspectannotation string) error {
+
+	valid_values := []string{"disabled", ""}
+
+	if !stringInSlice(valid_values, inspectannotation) {
+		return fmt.Errorf("invalid value for Inspect Annotation")
+	}
+
+	return nil
+}
+
+func validRebootAnnotation(rebootannotation string) error {
+	if rebootannotation == "" {
+		return nil
+	}
+
+	objStatus := &RebootAnnotationArguments{}
+	err := json.Unmarshal([]byte(rebootannotation), objStatus)
+	if err != nil {
+		return fmt.Errorf("failed to fetch Reboot from annotation, %w", err)
+	}
+
+	supported_modes := []string{"hard", "soft", ""}
+
+	if !stringInSlice(supported_modes, string(objStatus.Mode)) {
+		return fmt.Errorf("invalid RebootMode in RebootAnnotation")
+	}
+
+	return nil
+}
+
+func unmarshalStatusAnnotation(content []byte) (*BareMetalHostStatus, error) {
+	objStatus := &BareMetalHostStatus{}
+	if err := json.Unmarshal(content, objStatus); err != nil {
+		return nil, errors.Wrap(err, "Failed to fetch Status from annotation")
+	}
+	return objStatus, nil
+}
+
+func stringInSlice(strlist []string, str string) bool {
```

Comment:
We are using go 1.19 in go.mod, so in theory we could now use https://pkg.go.dev/golang.org/x/exp/slices#Contains for this.